### PR TITLE
#2 Reader as input for Parser

### PR DESCRIPTION
**Pull request by sentyaev**
I've add the ability to pass Reader to Parser.

Sorry for 'throws IOException', I'm a .NET developer, 
just don't know how you (Java developers) handle this.

Please help me handle this IOException in Java way, and I'll clean it up and will add tests.

Please do not be rude, I'm a .NET developer, but just starting with Java)

### CODE REVIEW

File: commonmark/src/main/java/org/commonmark/internal/DocumentParser.java [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
 import org.commonmark.internal.util.Parsing;
-import org.commonmark.internal.util.Substring;
 import org.commonmark.node.*;
 import org.commonmark.parser.block.*;
 
 import java.util.*;
 
-public class DocumentParser implements ParserState {
-
+public abstract class DocumentParser<T> implements ParserState {
```
Comment (robinst) [inline commonmark/src/main/java/org/commonmark/internal/DocumentParser.java]: I think adding a class hierarchy for this is a bit too complicated. I'd prefer to just have another method on `DocumentParser` that takes a `Reader`. The code before `incorporateLine` can be moved into the constructor, the part after into a single method that needs to be called by both `parse` methods.

File: commonmark/src/main/java/org/commonmark/internal/DocumentParser.java [hunk 2]
```diff
@@ -63,31 +63,19 @@ public DocumentParser(List<BlockParserFactory> blockParserFactories, InlineParse
...
-
+        
+        incorporateLines(input);
+        
         finalizeBlocks(activeBlockParsers, lineNumber);
         this.processInlines();
         return documentBlockParser.getBlock();
     }
-
+
```
Comment (robinst) [inline commonmark/src/main/java/org/commonmark/internal/DocumentParser.java]: What editor are you using? It shouldn't add unnecessary spaces on empty lines.

File: commonmark/src/main/java/org/commonmark/internal/ReaderParser.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+public class ReaderParser extends DocumentParser<Reader> {
+
+    public ReaderParser(List<BlockParserFactory> blockParserFactories, InlineParserImpl inlineParser) {
+        super(blockParserFactories, inlineParser);
+    }
+    
+    @Override
+    protected void incorporateLines(Reader input) throws IOException {
+        try (BufferedReader r = new BufferedReader(input)){
```
Comment (robinst) [inline commonmark/src/main/java/org/commonmark/internal/ReaderParser.java]: This causes the passed `input` to be closed at the end of the block, which is not what we want. We want the caller to handle this, we will just read from it until we're done. The typical caller then looks like this:

```java
try (BufferedReader reader = Files.newBufferedReader(filePath)) {
    Node node = parser.parse(reader);
    /* ... */
}
```

Also, I think we should do the following:

```java
BufferedReader bufferedReader;
if (input instanceof BufferedReader) {
    bufferedReader = (BufferedReader) input;
} else {
    bufferedReader = new BufferedReader(input);
}
```

File: commonmark/src/main/java/org/commonmark/internal/ReaderParser.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+            String line;
+            while((line = r.readLine()) != null) {
```
Comment (robinst) [inline commonmark/src/main/java/org/commonmark/internal/ReaderParser.java]: Missing space after `while`.

File: commonmark/src/main/java/org/commonmark/parser/Parser.java [hunk 5]
```diff
@@ -39,16 +43,26 @@ public static Builder builder() {
      * @param input the text to parse
      * @return the root node
      */
-    public Node parse(String input) {
+    public Node parse(String input) throws IOException {
```
Comment (robinst) [inline commonmark/src/main/java/org/commonmark/parser/Parser.java]: The `throws IOException` is not really necessary, see my comments above.

File: commonmark/src/test/java/org/commonmark/test/UsageExampleTest.java [hunk 6]
```diff
@@ -10,11 +12,20 @@
...
         Parser parser = Parser.builder().build();
         Node document = parser.parse("This is *Sparta*");
         HtmlRenderer renderer = HtmlRenderer.builder().escapeHtml(true).build();
         assertEquals("<p>This is <em>Sparta</em></p>\n", renderer.render(document));
     }
+    
+    @Test
+    public void two() throws IOException {
+        Parser parser = Parser.builder().build();
+        StringReader sr = new StringReader("This is *Sparta*");
```
Comment (robinst) [inline commonmark/src/test/java/org/commonmark/test/UsageExampleTest.java]: Let's add a test that uses a real file, e.g. `spec.txt`. You can do `UsageExampleTest.class..getResourceAsStream("/spec.txt")` to get an InputStream for the file. Compare it to rendering the same file using a `String` as an input.

Also, this class is for the examples that are used in `README`. Could you move it to a new `ParserTest` instead?

**Comment by robinst** [comment]
Thanks for the pull request :). See my inline comments.

Also, you will have to sign the following Contributor License Agreement (CLA) before I can accept the pull request (I'm sorry, but this is currently required):

https://na2.docusign.net/Member/PowerFormSigning.aspx?PowerFormId=3f94fbdc-2fbe-46ac-b14c-5d152700ae5d

File: commonmark/src/main/java/org/commonmark/internal/DocumentParser.java (hunk 2 above)
Comment (sentyaev) [inline commonmark/src/main/java/org/commonmark/internal/DocumentParser.java]: It's not my editor, it's me. I think I do not check for empty line, or just miss it.

File: commonmark/src/test/java/org/commonmark/test/UsageExampleTest.java (hunk 6 above)
Comment (sentyaev) [inline commonmark/src/test/java/org/commonmark/test/UsageExampleTest.java]: sure

**Comment by sentyaev** [comment]
I'm sign form you provide.

Thanks for reply.

Your comments are reasonable.
So updates will come soon.

**Pull request closed without merging**